Tests/ASLayoutEngineTests.mm

```diff
@@ -0,0 +1,521 @@
+  [nodeA transitionLayoutWithAnimation:NO shouldMeasureAsync:YES measurementCompletion:^{
+    measurementCompletionBlock();
+  }];
+
+  // This block will get run after bg layout calculate starts, but before measurementCompletion
+  __block BOOL injectedMainThreadWorkDone = NO;
+  injectedMainThreadWork = ^{
+    injectedMainThreadWorkDone = YES;
+
+    // BUG: nodeC survives but gets a layout update during transition.
+    // When this layout pass comes down, its parent will set its new frame which will
+    // trigger the layout pass to go down into it and we will measure against that new bounds.
+    // The new bounds will be the post-transition size.
+    if (!enforceCorrectBehavior) {
+      auto postTransitionSize = [fixture2 layoutForNode:nodeC].size;
+      OCMExpect([nodeC.mock calculateLayoutThatFits:ASSizeRangeMake(postTransitionSize)]).onMainThread();
+    }
+
+    [window layoutIfNeeded];
+
+    // Ensure we're still on the old layout. We should stay on this until the transition completes.
+    [self verifyFixture:fixture1];
+  };
+
+  measurementCompletionBlock = ^{
+    XCTAssert(injectedMainThreadWorkDone, @"We hoped to get onto the main thread before the measurementCompletion callback ran.");
+  };
+
+  for (ASLayoutTestNode *node in allNodes) {
+    OCMVerifyAllWithDelay(node.mock, verifyDelay);
+  }
+
+  [self verifyFixture:fixture2];
+}
+
+/**
+ * Start at fixture 1.
+ * Trigger an async transition to fixture 2.
+ * While it's measuring, on main switch to fixture 4 (setNeedsLayout A, D) and run a CA layout pass.
+ *
+ * Correct behavior, we end up at fixture 4 since it's newer.
+ * Current incorrect behavior, we end up at fixture 2 and we remeasure surviving node C.
+ */
+- (void)testASetNeedsLayoutInterferingWithTheCurrentTransition
+{
+  static BOOL enforceCorrectBehavior = NO;
+
+  [self stubCalculatedLayoutDidChange];
+  [self runFirstLayoutPassWithFixture:fixture1];
+
+  [fixture2 apply];
+
+  // Expect A, C, E to calculate new layouts off-main
+  // dispatch_once onto main to run our injectedMainThread work while the transition calculates.
+  __block dispatch_block_t injectedMainThreadWork = nil;
+  for (ASLayoutTestNode *node in @[ nodeA, nodeC, nodeE ]) {
+    [fixture2 withSizeRangesForNode:node block:^(ASSizeRange sizeRange) {
+      OCMExpect([node.mock calculateLayoutThatFits:sizeRange])
+      .offMainThread()
+      .andDo(^(NSInvocation *inv) {
+        // On first calculateLayoutThatFits, schedule our injected main thread work.
+        static dispatch_once_t onceToken;
+        dispatch_once(&onceToken, ^{
+          dispatch_async(dispatch_get_main_queue(), ^{
+            injectedMainThreadWork();
+          });
+        });
+      });
+    }];
+  }
+
+  // The code in this section is designed to move in time order, all on the main thread:
+
+  OCMExpect([nodeA.mock animateLayoutTransition:OCMOCK_ANY]).onMainThread();
+  OCMExpect([nodeA.mock didCompleteLayoutTransition:OCMOCK_ANY]).onMainThread();
```

Comment:
These method calls are only expected when the correct behavior is **not** enforced.  The correct behavior is that the layout transition will be cancelled and this it won't animate and complete.